src/envs/pybullet_blocks.py

```diff
@@ -0,0 +1,728 @@
+                      params: Array) -> bool:
+            del memory, objects, params  # unused
+            current_val = state.get(self._robot, "fingers")
+            dist = (target_val - current_val)**2
+            return dist < self._grasp_tol
+
+        return ParameterizedOption(name,
+                                   types=types,
+                                   params_space=params_space,
+                                   policy=_policy,
+                                   initiable=lambda _1, _2, _3, _4: True,
+                                   terminal=_terminal)
+
+    def _move_relative_to_table(self, name: str,
+                                rel_or_abs_target_pose: Pose3D,
+                                rel_or_abs: Tuple[str, str,
+                                                  str], types: Sequence[Type],
+                                params_space: Box) -> ParameterizedOption:
+        """Creates a ParameterizedOption for moving to a target pose relative
+        to the table.
+
+        Each of the three dimensions of the target pose can be specified
+        relatively or absolutely. The rel_or_abs argument indicates
+        whether each dimension is relative ("rel") or absolute ("abs").
+        """
+
+        def _get_current_and_target_pose(
+                state: State, objects: Sequence[Object],
+                params: Array) -> Tuple[Pose3D, Pose3D]:
+            robot, = objects
+            # De-normalize parameters to actual table coordinates.
+            x_norm, y_norm = params
+            x = self.x_lb + (self.x_ub - self.x_lb) * x_norm
+            y = self.y_lb + (self.y_ub - self.y_lb) * y_norm
+            current_pose = (state[robot][0], state[robot][1], state[robot][2])
+            table_pose = (x, y, self.table_height)
+            target_pose = self._convert_rel_abs_to_abs(table_pose,
+                                                       rel_or_abs_target_pose,
+                                                       rel_or_abs)
+            return current_pose, target_pose
+
+        return self._move_to_pose(name, types, params_space,
+                                  _get_current_and_target_pose)
+
+    def _get_move_toward_waypoint_action(self, gripper_pose: Pose3D,
+                                         target_pose: Pose3D) -> Action:
+        action = self._move_gain * np.subtract(target_pose, gripper_pose)
+        action_norm = np.linalg.norm(action)  # type: ignore
+        if action_norm > self._max_vel_norm:
+            action = action * self._max_vel_norm / action_norm
+        action = np.r_[action, 0.0]
+        action = np.clip(action, self.action_space.low, self.action_space.high)
+        return Action(action.astype(np.float32))
+
+    @staticmethod
+    def _convert_rel_abs_to_abs(current_pose: Pose3D, rel_or_abs_pose: Pose3D,
+                                rel_or_abs: Tuple[str, str, str]) -> Pose3D:
+        abs_pos = []
+        for i in range(3):
```

Comment:
range(len(current_pose)), and assert that `len(current_pose) == len(rel_or_abs_pose) == len(rel_or_abs)`